File: C/kuramoto.c
#include <math.h>   // for maths functions
#include <stdlib.h> // for malloc, etc.
#include <stdio.h> // for malloc, etc.

// NOTE:  C is row-major; bear in mind when writing interfaces! E.g. for
// Matlab (column-major) you should transpose the matrices K and a before calling.

void kuramoto_euler	// Euler method
(
	const size_t        N,  // number of oscillators
	const size_t        n,  // number of integration increments
	const double* const w,  // dt*frequencies
	const double* const K,  // dt*(coupling constants)
	double*       const h   // oscillator phases, to be computed by numerical ODE (pre-initialised with input)
)
{
	for (double* ht=h; ht<h+N*(n-1); ht+=N) {
		double* const ht1 = ht+N;
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double hti = ht[i];
			double dhti = w[i];
			for (size_t j=0; j<N; ++j) dhti += Ki[j]*sin(ht[j]-hti);
			ht1[i] += hti+dhti; // update next time step (adding in input already in ht1)
		}
	}
}

void kuramoto_eulerpl // Euler method with phase lags
(
	const size_t        N,  // number of oscillators
	const size_t        n,  // number of integration increments
	const double* const w,  // dt*frequencies
	const double* const K,  // dt*(coupling constants)
	const double*       a,  // phase lags
	double*       const h   // oscillator phases, to be computed by numerical ODE (pre-initialised with input)
)
{
	for (double* ht=h; ht<h+N*(n-1); ht+=N) {
		double* const ht1 = ht+N;
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double* const ai = a+N*i;
			const double hti = ht[i];
			double dhti = w[i];
			for (size_t j=0; j<N; ++j) dhti += Ki[j]*sin(ht[j]-hti-ai[j]);
			ht1[i] += hti+dhti; // update next time step (adding in input already in ht1)
		}
	}
}

void kuramoto_rk4 // Classic Runge-Kutta (RK4)
(
	const size_t        N, // number of oscillators
	const size_t        n, // number of integration increments
	const double* const w, // dt*frequencies
	const double* const K, // dt*(coupling constants)
	double*       const h, // oscillator phases, to be computed by numerical ODE (pre-initialised with input)
	double*             k1 // buffer for RK4 coefficients (size must be 4*N)
)
{
	double* const k2 = k1+N;
	double* const k3 = k2+N;
	double* const k4 = k3+N;

	for (double* ht=h; ht<h+N*(n-1); ht+=N) {

		// k1
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double hti = ht[i];
			double ki = w[i];
			for (size_t j=0; j<N; ++j) ki += Ki[j]*sin(ht[j]-hti);
			k1[i] = ki;
		}

		// k2
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double htk1i = ht[i]+k1[i];
			double ki = w[i];
			for (size_t j=0; j<N; ++j) ki += Ki[j]*sin(ht[j]+k1[j]-htk1i);
			k2[i] = ki/2.0;
		}

		// k3
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double htk2i = ht[i]+k2[i];
			double ki = w[i];
			for (size_t j=0; j<N; ++j) ki += Ki[j]*sin(ht[j]+k2[j]-htk2i);
			k3[i] = ki/2.0;
		}

		// k4
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double htk3i = ht[i]+k3[i];
			double ki = w[i];
			for (size_t j=0; j<N; ++j) ki += Ki[j]*sin(ht[j]+k3[j]-htk3i);
			k4[i] = ki;
		}

		// update next time step (adding in input already in ht1)
		double* const ht1 = ht+N;
		for (size_t i=0; i<N; ++i) {
			ht1[i] += ht[i] + (k1[i]+4.0*k2[i]+4.0*k3[i]+k4[i])/6.0;
		}
	}
}

void kuramoto_rk4pl // Classic Runge-Kutta (RK4) with phase lags
(
	const size_t        N, // number of oscillators
	const size_t        n, // number of integration increments
	const double* const w, // dt*frequencies
	const double* const K, // dt*(coupling constants)
	const double*       a, // phase lags
	double*       const h, // oscillator phases, to be computed by numerical ODE (pre-initialised with input)
	double*             k1 // buffer for RK4 coefficients (size must be 4*N)
)
{
	double* const k2 = k1+N;
	double* const k3 = k2+N;
	double* const k4 = k3+N;

	for (double* ht=h; ht<h+N*(n-1); ht+=N) {

		// k1
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double* const ai = a+N*i;
			const double hti = ht[i];
			double ki = w[i];
			for (size_t j=0; j<N; ++j) ki += Ki[j]*sin(ht[j]-hti-ai[j]);
			k1[i] = ki;
		}

		// k2
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double* const ai = a+N*i;
			const double htk1i = ht[i]+k1[i];
			double ki = w[i];
			for (size_t j=0; j<N; ++j) ki += Ki[j]*sin(ht[j]+k1[j]-htk1i-ai[j]);
			k2[i] = ki/2.0;
		}

		// k3
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double* const ai = a+N*i;
			const double htk2i = ht[i]+k2[i];
			double ki = w[i];
			for (size_t j=0; j<N; ++j) ki += Ki[j]*sin(ht[j]+k2[j]-htk2i-ai[j]);
			k3[i] = ki/2.0;
		}

		// k4
		for (size_t i=0; i<N; ++i) {
			const double* const Ki = K+N*i;
			const double* const ai = a+N*i;
			const double htk3i = ht[i]+k3[i];
			double ki = w[i];
			for (size_t j=0; j<N; ++j) ki += Ki[j]*sin(ht[j]+k3[j]-htk3i-ai[j]);
			k4[i] = ki;
		}

		// update next time step (adding in input already in ht1)
		double* const ht1 = ht+N;
		for (size_t i=0; i<N; ++i) {
			ht1[i] += ht[i] + (k1[i]+4.0*k2[i]+4.0*k3[i]+k4[i])/6.0;
		}
	}
}

void order_param // calculate order parameter magnitude
(
	const size_t N,
	const size_t n,
	const double* const h,
	double* const r
)
{
	const double OON = 1.0/(double)N;
	double* rt = r;
	for (const double* ht=h; ht<h+N*n; ht+=N,++rt) {
		double cmean = 0.0;
		double smean = 0.0;
		for (size_t i=0; i<N; ++i) {
#ifdef _GNU_SOURCE
			double c,s;
			sincos(ht[i],&s,&c);
			cmean += c;
			smean += s;
#else
			cmean += cos(ht[i]);
			smean += sin(ht[i]);
#endif
		}
		cmean *= OON;
		smean *= OON;
		*rt = hypot(cmean,smean);
	}
}

static inline double wmpi2pi(const double x) // wrap to [-pi,pi)
{
	return x > 0.0 ? fmod(x+M_PI,2.0*M_PI)-M_PI : fmod(x-M_PI,2.0*M_PI)+M_PI;
}

void phase_wrap // wrap all oscillator phases
(
	const size_t N,
	const size_t n,
	const double* const h,
	double* const theta
)
{
	for (size_t k=0; k<N*n; ++k) {
		theta[k] = wmpi2pi(h[k]);
	}
}

File: Matlab/kuramoto.m
function [h,r,psi] = kuramoto(N,n,dt,w,K,a,h0,I,mode)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Wrapper for kuramoto C mex functions - all input checking happens here!
%
% To compile "mex" files, see Makefile in this directory
%
% N     number of oscillators                 (positive integer)
% n     number of time increments             (positive integer)
% dt    time integration step                 (scalar)                             : seconds
% w     oscillator frequencies                (scalar or vector of length N)       : radians/second
% K     oscillator coupling constants         (scalar or square matrix of size N)  : radians/second
% a     phase lags                            (scalar or square matrix of size N)  : radians
% h0    initial phases                        (vector of length N)                 : radians
% I     input noise                           (n x N matrix or empty for no input) : radians/sqrt(second)
% mode  simulation mode                       ('Euler' or 'RK4')                   : string
%
% h     oscillator phases (unwrapped)         (N x n matrix)                       : radians
% r     order parameter magnitude             (row vector of length n)             : dimensionless
% psi   order parameter phase (wrapped)       (row vector of length n)             : radians
%
% NOTE 1: Phases are interpreted as radians on [0,2*pi), and frequencies angular; i.e., in radians/second
%
% NOTE 2: K(i,j) is connection strength from oscillator j to oscillator i.
%
% NOTE 3: Euler method is faster (by a factor of about 5), but RK4 is more accurate.
%
% NOTE 4: To wrap the oscillator phases h to [-pi,pi), do:
%
%     h = mod(h+pi,2*pi)-pi;
%
% See kuramoto_demo.m script for example usage.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Input checks - essential, as the mex functions do NO error checking!

N = double(N);
assert(isscalar(N) && floor(N) == N && N > 0,'Number of oscillators must be a positive scalar integer');

n = double(n);
assert(isscalar(n) && floor(n) == n && n > 0,'Number of time increments must be a positive scalar integer');

assert(isa(dt,'double') && isscalar(dt) && dt > 0,'Integration increment must be a positive scalar double');

assert(isa(w,'double'),'Frequencies must be a scalar double or a vector of doubles matching the specified number of oscillators');
if isscalar(w)
	w = w*ones(N,1);
else
	assert(isvector(w) && length(w) == N,'Frequencies must be a scalar double or a vector of doubles matching the specified number of oscillators');
end

assert(isa(K,'double'),'Coupling constants must be a scalar double, or a square matrix of doubles matching the specified number of oscillators');
if isscalar(K)
	K = K*ones(N);    % uniform coupling
	K(1:N+1:N*N) = 0; % zeros on diagonal!
else
	assert(ismatrix(K) && size(K,1) == N && size(K,2) == N,'Coupling constants must be a scalar double, or a square matrix of doubles matching the specified number of oscillators');
end

if ~isempty(a)
	assert(isa(a,'double'),'Phase lags must be empty, a scalar double, or a square matrix of doubles matching matching the specified number of oscillators');
	if isscalar(a)
		a = a*ones(N);    % uniform phase lag
		a(1:N+1:N*N) = 0; % zeros on diagonal!
	else
		assert(ismatrix(a) && size(a,1) == N && size(a,2) == N,'Phase lags must be empty, a scalar double, or a square matrix of doubles matching the specified number of oscillators');
	end
end

assert(isempty(h0) || (isa(h0,'double') && isvector(h0) && length(h0) == N),'Initial phases must be empty, or a vector of doubles of length N');

assert(isempty(I) || (isa(I,'double') && ismatrix(I) && size(I,1) == n && size(I,2) == N),'Input noise must be empty, or an n x N matrix of doubles');

if isempty(mode)
	RK4 = 1;
else
	assert(ischar(mode),'Simulation mode must be empty, or ''Euler'' or ''RK4''');
	switch upper(mode)
		case 'RK4',   RK4 = 1;
		case 'EULER', RK4 = 0;
		otherwise,    error('Unknown simulation mode; must be empty, ''Euler'' or ''RK4''');
	end
end

% Call mex ODE simulation (returned phase matrix h is n x N)
%
% We transpose K so that K(i,j) is connection strength j --> i, same for a
%
% Input noise is Weiner (Brownian), so scaled by sqrt(dt); cf. Ito simulation of Ornstein-Uhlenbeck process

h = kuramoto_mex(N,n,w*dt,K'*dt,a',h0,I*sqrt(dt),RK4);

% Order parameter (if requested)

if nargout > 1
	x = mean(cos(h));
	y = mean(sin(h));
	r = hypot(x,y);
	if nargout > 2
		psi = atan2(y,x); % wrapped on [-pi,pi)
	end
end
